### DISCUSSION TITLE
Feature: weakref-based cleanup

### DISCUSSION THREAD

**Discussion by michaelfig** [Agoric/agoric-sdk#5477] [Ideas]
Proposed by @erights as an optional feature.

**Comment by zarutian** [comment]
https://tc39.es/proposal-weakrefs/ is of intrest to those issue readers who do not know how that would work.

In short. A FinalizationRegistry would be used to keep track of if presences local to a vat yet pointing into remote vats have been locally gc'ed. If so we can inform the remote vat of our disintrest into the object that had been referenced.
(Do not ask about object graph cycles of garbage that span multiple vats. That problem is not only thorny, it is an entire briar patch. (self-memo: staleness-colouring via vat-vectorclock-timestamps))

**Reply by michaelfig** [reply]
Thank you for the pointer, @zarutian!

I've implemented this feature in `@endo/captp` in a pending PR, @endojs/endo#1205.

**Reply by michaelfig** [reply]
... and @warner already did it some time ago for SwingSet-based vats.
